### PR TITLE
[WIP] Get endpoints for service instead of pods

### LINKED ISSUE #9: ES cluster cannot startup
Following the deployment of an app with 2GB of memory usage, ES cannot be resized or deployed in k8s. 

Following error when spinning up ES cluster in k8s: 

```bash
2015-04-25T00:05:01.856141447Z Caused by: com.fasterxml.jackson.databind.JsonMappingException: Numeric value (2147483648) out of range of int
2015-04-25T00:05:01.856141447Z  at [Source: sun.net.www.protocol.http.HttpURLConnection$HttpInputStream@511e7433; line: 2062, column: 35] (through reference chain: io.fabric8.kubernetes.api.model.PodList["items"]->io.fabric8.kubernetes.api.model.Pod["desiredState"]->io.fabric8.kubernetes.api.model.PodState["manifest"]->io.fabric8.kubernetes.api.model.ContainerManifest["containers"]->io.fabric8.kubernetes.api.model.Container["memory"])
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:232)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.JsonMappingException.wrapWithPath(JsonMappingException.java:197)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.wrapAndThrow(BeanDeserializerBase.java:1415)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:244)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:118)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:232)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:206)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:25)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:538)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:99)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:242)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:118)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:538)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:99)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:242)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:118)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:538)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:99)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:242)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:118)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:232)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:206)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:25)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:538)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:99)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:242)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:118)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.ObjectReader._bind(ObjectReader.java:1232)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:676)
2015-04-25T00:05:01.856141447Z 	at com.fasterxml.jackson.jaxrs.base.ProviderBase.readFrom(ProviderBase.java:800)
2015-04-25T00:05:01.856141447Z 	at org.apache.cxf.jaxrs.utils.JAXRSUtils.readFromMessageBodyReader(JAXRSUtils.java:1322)
2015-04-25T00:05:01.856141447Z 	at org.apache.cxf.jaxrs.impl.ResponseImpl.doReadEntity(ResponseImpl.java:369)
2015-04-25T00:05:01.856141447Z 	... 15 more
```

### PR DESCRIPTION
**Pull request by pires**
This is work-in-progress to fix #9 and improve node-discovery efficiency.

* Need guidance on how to test this
* Need [a new release of fabric8](https://github.com/fabric8io/fabric8/issues/3817)

### CODE REVIEW

**Comment by pires** [comment]
@jimmidyson sorry about the formatting changes and some cosmetic changes but I tried hard to keep up  with your conventions. How does this look?

I'm going to need you help testing this with perhaps a snapshot release or something? I'm open to ideas.

File: src/main/java/io/fabric8/elasticsearch/discovery/k8s/K8sUnicastHostsProvider.java [hunk 1]
```diff
@@ -119,67 +115,44 @@ public K8sUnicastHostsProvider(Settings settings,
...
-
-                                logger.trace("adding {}, address {}, transport_address {}, status {}", pod.getId(),
-                                        podIp, addresses[0], status);
-                                cachedDiscoNodes.add(new DiscoveryNode("#cloud-" + pod.getId() + "-" + 0, addresses[0], Version.CURRENT));
-                            }
-                        }
+                    // pod IP is a single IP Address. We need to build a TransportAddress from it
+                    final TransportAddress[] addresses;
+                    try {
+                        addresses = transportService.addressesFromString(endpointIp.concat(":9300"));
```
Comment (jimmidyson) [inline src/main/java/io/fabric8/elasticsearch/discovery/k8s/K8sUnicastHostsProvider.java]: Is there no way to get the port from the endpoint list?

**Comment by jimmidyson** [comment]
@pires Thanks for this. Overall looks good. I wonder if we could use fabric8 arquillian to add some integration tests for this? Happy to have that later though. For now you should be able to build locally & use in a local docker image? Don't like publishing snapshots personally & would rather not do that.

File: src/main/java/io/fabric8/elasticsearch/discovery/k8s/K8sUnicastHostsProvider.java (hunk 1 above)
Comment (jimmidyson) [inline src/main/java/io/fabric8/elasticsearch/discovery/k8s/K8sUnicastHostsProvider.java]: Yes this should be doable. See https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/types.go#L1091 ff for more details. Basically it's a Cartesian product between addresses and ports in each endpoint subset.

Comment (pires) [inline src/main/java/io/fabric8/elasticsearch/discovery/k8s/K8sUnicastHostsProvider.java]: While I'm hardcoding port 9300 and thus limiting the user choices here, I'm concerned that a multiport service will make it harder for us to determine which port to use.

**Comment by pires** [comment]
> I wonder if we could use fabric8 arquillian to add some integration tests for this?

Probably would be a good thing to do. Would we need something like `jube` to actually bring Kubernetes up?

> For now you should be able to build locally & use in a local docker image? Don't like publishing snapshots personally & would rather not do that.

OK. Will do that tomorrow then.

File: src/main/java/io/fabric8/elasticsearch/discovery/k8s/K8sUnicastHostsProvider.java (hunk 1 above)
Comment (jimmidyson) [inline src/main/java/io/fabric8/elasticsearch/discovery/k8s/K8sUnicastHostsProvider.java]: Good point. I think using port 9300 for now is fine. Almost all deployments will use that I'm sure.

**Comment by jimmidyson** [comment]
No - fabric8 arquillian can spin up pods in a kubernetes cluster for integration testing. See http://fabric8.io/guide/testing.html for more details. It's pretty awesome! We're using it in some CD scenarios & it works great. But like I said, this can come later - we will merge this once you're happy with it.

Let me know how your testing goes.

**Comment by mulloymorrow** [comment]
+1

**Comment by pires** [comment]
I'm sorry, I have been overwhelmed with unrelated work.

>  For now you should be able to build locally & use in a local docker image? 

Yes, will do this for now.
Unfortunately, I think I won't be able to look at Arquillian in the near future.

**Comment by mulloymorrow** [comment]
I'll try and look at Arquillian and test this out locally.

**Comment by pires** [comment]
@mulloymorrow that would be A-W-E-S-O-M-E!

**Comment by mulloymorrow** [comment]
Can we merge this pull request with a new branch other than master? I'd like to include these commits in a fork for testing.

**Comment by jimmidyson** [comment]
@mullowmorrow Perhaps @pires could add you as a collaborator on his fork?

File: src/main/java/io/fabric8/elasticsearch/discovery/k8s/K8sUnicastHostsProvider.java [hunk 2]
```diff
@@ -119,67 +115,44 @@ public K8sUnicastHostsProvider(Settings settings,
...
-
-            try {
-                String podIp = pod.getCurrentState().getPodIP();
-
-                if (podIp.equals(currentIpAddress)) {
-                    // We found the current node.
-                    // We can ignore it in the list of DiscoveryNode
-                    logger.trace("current node found. Ignoring {} - {}", pod.getId(), podIp);
+        // get all endpoints for service
+        final Endpoints endpoints = getKubernetes().endpointsForService(this.selector, null);
```
Comment (mulloymorrow) [inline src/main/java/io/fabric8/elasticsearch/discovery/k8s/K8sUnicastHostsProvider.java]: IAE thrown by kubernetes-api-2.1.0::KubernetesClient.validateNamespace: Template parameter set to null. 

```bash
2015-05-28T18:17:16.071899300Z [2015-05-28 18:17:16,071][WARN ][io.fabric8.elasticsearch.discovery.k8s.K8sUnicastHostsProvider] [Cayman] Exception caught during discovery java.lang.IllegalArgumentException : Template parameter value is set to null
2015-05-28T18:17:16.071899300Z java.lang.IllegalArgumentException: Template parameter value is set to null
2015-05-28T18:17:16.071899300Z 	at org.apache.cxf.jaxrs.impl.UriBuilderImpl.doBuild(UriBuilderImpl.java:97)
2015-05-28T18:17:16.071899300Z 	at org.apache.cxf.jaxrs.impl.UriBuilderImpl.buildFromEncoded(UriBuilderImpl.java:230)
2015-05-28T18:17:16.071899300Z 	at org.apache.cxf.jaxrs.client.ClientProxyImpl.invoke(ClientProxyImpl.java:180)
2015-05-28T18:17:16.071899300Z 	at com.sun.proxy.$Proxy29.endpointsForService(Unknown Source)
2015-05-28T18:17:16.071899300Z 	at io.fabric8.elasticsearch.discovery.k8s.K8sUnicastHostsProvider.getNodesFromKubernetesSelector(K8sUnicastHostsProvider.java:119)
2015-05-28T18:17:16.071899300Z 	at io.fabric8.elasticsearch.discovery.k8s.K8sUnicastHostsProvider.buildDynamicNodes(K8sUnicastHostsProvider.java:102)
2015-05-28T18:17:16.071899300Z 	at org.elasticsearch.discovery.zen.ping.unicast.UnicastZenPing.sendPings(UnicastZenPing.java:313)
2015-05-28T18:17:16.071899300Z 	at org.elasticsearch.discovery.zen.ping.unicast.UnicastZenPing$2.doRun(UnicastZenPing.java:228)
2015-05-28T18:17:16.071899300Z 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:36)
2015-05-28T18:17:16.071899300Z 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
2015-05-28T18:17:16.071899300Z 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
2015-05-28T18:17:16.071899300Z 	at java.lang.Thread.run(Thread.java:745)
```

Comment (pires) [inline src/main/java/io/fabric8/elasticsearch/discovery/k8s/K8sUnicastHostsProvider.java]: We probably need to set a default namespace, like `default`.

Comment (mulloymorrow) [inline src/main/java/io/fabric8/elasticsearch/discovery/k8s/K8sUnicastHostsProvider.java]: Testing `default`

**Comment by jimmidyson** [comment]
ok to test

**Comment by jimmidyson** [comment]
retest this please

**Comment by jimmidyson** [comment]
@pires @mulloymorrow I've merged this locally & made some changes to support your use case, getting endpoints from service name. I've tested & it works fine. Please can you review `service-endpoints` branch? If you're happy then I will merge this to master & release.

**Comment by mulloymorrow** [comment]
We're running in v1beta2, currently. Services in our environment can be found vi `/api/v1beta2/services/<SERVICE_NAME>?namespace=default`.

**Comment by jimmidyson** [comment]
As v1beta1 & v1beta2 has been dropped from kubernetes we've removed those api versions from the fabric8 client, only supporting v1beta3 ATM. Soon that will change to only v1. Sorry but you'll need to upgrade your cluster in order to use this plugin going forward. We're targeting the v1 API in line with kubernetes 1.0 release.

**Comment by pires** [comment]
I agree with Jimmi. You really need to upgrade.

**Comment by mulloymorrow** [comment]
Switched over to Gcloud! Will be testing there.

**Comment by pires** [comment]
You mean GKE?

Also, I'm sorry for not updating this, but right now there's a lot going on regarding reading the Kubernetes API that will have an impact on a lot of the stuff I'm working on and most probably this project as well.

@jimmidyson did you check the `elasticsearch` example from @satnam6502 on the Kubernetes tree? They're using service-accounts for allowing secure access to the API and it seems read-only on an unsecure port will be dropped soon.

**Comment by mulloymorrow** [comment]
@pires , Yes. The Google Container Engine.

**Comment by jimmidyson** [comment]
@pires I'm working on the fabric8 kubernetes API support for service accounts now, but that shouldn't stop this PR going through. Separate issue for that please.

**Comment by pires** [comment]
@jimmidyson point taken `service-endpoints` branch works for me with Kubernetes 0.17.0 and 0.18.0, so looks good. Care to merge then?

**Comment by jimmidyson** [comment]
@pires Cool thanks for the feedback. Btw seen the advice to use SRV records rather than API server calls?

**Comment by pires** [comment]
@jimmidyson yes, but see https://github.com/GoogleCloudPlatform/kubernetes/issues/9107. We should be able to work without DNS, because it's not something that is granted in every Kubernetes cluster.

**Comment by jimmidyson** [comment]
@pires Agreed - was just pointing out that I prefer DNS if available.

**Comment by jimmidyson** [comment]
Superseded by #12

**Comment by jimmidyson** [comment]
Thanks @pires!

**Comment by pires** [comment]
Thank you @jimmidyson

**Comment by mulloymorrow** [comment]
Thanks @pires @jimmidyson

**Pull request closed without merging**